dowhy/gcm/influence.py

```diff
@@ -376,3 +431,48 @@ def icc_set_function(subset: np.ndarray) -> Union[np.ndarray, float]:
         return attribution_func(shape_into_2d(predictions), target_values)
 
     return estimate_shapley_values(icc_set_function, noise_samples.shape[1], shapley_config)
+
+
+def _get_icc_noise_function(
+    causal_model,
+    target_node,
+    prediction_model,
+    noise_samples,
+    node_names,
+    target_samples,
+    auto_assign_quality,
+    target_is_categorical,
+) -> Callable[[np.ndarray], np.ndarray]:
+    if prediction_model == "approx":
+        prediction_model = auto.select_model(noise_samples, target_samples, auto_assign_quality)
+        prediction_model.fit(noise_samples, target_samples)
+
+        if target_is_categorical:
+            prediction_method = prediction_model.predict_probabilities
+        else:
+            prediction_method = prediction_model.predict
+    elif prediction_model == "exact":
+
+        def exact_model(X: np.ndarray) -> np.ndarray:
+            return compute_data_from_noise(causal_model, pd.DataFrame(X, columns=[x for x in node_names]))[
+                target_node
+            ].to_numpy()
+
+        if target_is_categorical:
+            list_of_classes = cast(ClassifierFCM, causal_model.causal_mechanism(target_node)).classifier_model.classes
+
+            def prediction_method(X):
+                return (shape_into_2d(exact_model(X)) == list_of_classes).astype(float)
+
+        else:
+            prediction_method = exact_model
+    elif isinstance(prediction_model, str):
+        raise ValueError(
+            "Invalid value for prediction_model: %s! This should either be an instance of a PredictionModel or"
+            "one of the two string options 'exact' or 'approx'." % prediction_model
+        )
+    else:
```

Comment:
else is not needed anymore

See also: [Why You Shouldn't Nest Your Code
 ](https://www.youtube.com/watch?v=CFRhGnuXG-4)

Oh, just saw: this code was actually just moved. Feel free to postpone my proposed refactoring. Either way you choose, I think it's worth doing it.